urllib3/_collections.py

```diff
@@ -98,6 +100,129 @@ def keys(self):
+    Key access is done through acquiring leases, which have to be released
+    afterwards. Inserting keys beyond ``maxsize`` throw away old keys, but not
+    before ensuring that all leases on such keys have been released.
+
+    Note: eviction/disposal is triggered on inserting new keys, not on
+    releasing of the last lease. In the case of connection pooling, this means
+    we will maintain the connections as long as possible.
+
+    This does not expose any means to clear the pool, because it would involve
+    waiting for all leases to be released. If there are concurrent acquisition
+    of leases, the clear method may block indefinitely.
+
+    :param maxsize:
+        Maximum number of recent elements to retain.
+
+    :param dispose_func:
+        Every time an item is evicted from the container,
+        ``dispose_func(value)`` is called.  Callback which will get called
+    """
+    ContainerCls = OrderedDict
+
+    def __init__(self, maxsize=10, dispose_func=None):
+        self._maxsize = maxsize
+        self._dispose_func = dispose_func
+
+        self._container = self.ContainerCls()
+        self.lock = RLock()
+
+        self._freelist = self.ContainerCls()
+        self._nwaiters = 0
+        self._container_not_full = Condition(self.lock)
+        self._victim = None
+
+    def acquire(self, key, constructor):
+        """
+        Acquires a lease on ``key`` inside this pool, if it has already
+        existed. If it hasn't, ``constructor`` is called to generate the value
+        and it is inserted into the pool beforehand.
+
+        If the insertion is done while the pool is full, another key is
+        evicted in a FIFO order, but not before all leases on the key have been
+        released.
+
+        :param key:
+            Key to acquire lease upon.
+
+        :param constructor:
+            Function-like, used to construct the new item to lease on if it
+            hasn't existed yet.
+
+        :returns value:
+            Value associated with key, or the value created by ``constructor``
+            if no such item can be found.
+        """
+        with self.lock:
+            result = self._container.get(key, _Null)
+
+            if result is not _Null:
+                refcount, value = self._container.pop(key)
+                self._container[key] = [refcount + 1, value]
+                _ = self._freelist.pop(key, _Null)
+                return value
+            else:
+                item = constructor()
+
+                if len(self._container) < self._maxsize:
+                    self._container[key] = [1, item]
+                else:
+                    evicted_value = _Null
+                    if len(self._freelist) > 0:
+                        _, evicted_value = self._freelist.popitem(last=False)
+                    else:
+                        self._nwaiters += 1
+                        self._container_not_full.wait()
```

Comment:
This doesn't work. The wait on the condition variable is done while holding `self.lock`. However, the *notify* of the condition variable requires holding `self.lock`, which means if anyone is waiting for a connection to be returned to this queue (in order to evict it) they will block all further queue activity forever. This is a pretty classic deadlock.

In this instance, I also think it's a bit overkill. There's no reason to block threads on this. Instead, we should just keep track of how many connection pools we have created. If that number goes above the "size" of this pool, then if there are any connections in the pool that have not been extracted they should be removed. If all connections from this pool have been removed, then they should be evicted in the order they're returned to the pool until the count drops below a certain level.

This means we don't block people forever waiting for connectionpools to be returned. It does allow a slight overcommitment of resources, but if people don't want that then they should probably just use a semaphore or something similar to manage how many outstanding requests they have.